bolero/optimizer/ccmaes.py

```diff
@@ -0,0 +1,338 @@
+        self.log_to_stdout = log_to_stdout
+        self.random_state = random_state
+
+    def init(self, n_params, n_context_dims):
+        """Initialize optimizer.
+
+        Parameters
+        ----------
+        n_params : int
+            number of parameters
+
+        n_context_dims : int
+            number of dimensions of the context space
+        """
+        self.logger = get_logger(self, self.log_to_file, self.log_to_stdout)
+
+        self.random_state = check_random_state(self.random_state)
+
+        self.it = 0
+
+        if self.initial_params is None:
+            self.initial_params = np.zeros(n_params)
+        else:
+            self.initial_params = np.asarray(self.initial_params).astype(
+                np.float64, copy=True)
+        if n_params != len(self.initial_params):
+            raise ValueError("Number of dimensions (%d) does not match "
+                             "number of initial parameters (%d)."
+                             % (n_params, len(self.initial_params)))
+        self.n_params = n_params
+        if self.covariance is None:
+            self.covariance = np.eye(n_params)
+        elif self.covariance.ndim == 1:
+            self.covariance = np.diag(self.covariance)
+
+        self.context = None
+        self.params = None
+        self.reward = None
+
+        self.policy_ = ContextTransformationPolicy(
+            LinearGaussianPolicy, n_params, n_context_dims,
+            context_transformation=self.context_features,
+            mean=self.initial_params, covariance_scale=1.0, gamma=self.gamma,
+            random_state=self.random_state)
+        self.cov = np.copy(self.covariance)
+        self.var = self.variance
+        self.policy_.policy.Sigma = self.variance * self.covariance
+
+        self.n_total_dims = n_params + n_context_dims
+        if self.n_samples_per_update is None:
+            self.n_samples_per_update = (
+                4 + int(3 * np.log(self.n_total_dims)) *
+                (1 + 2 * n_context_dims))
+
+        self.hsig_threshold = 2 + 4.0 / (self.n_params + 1)
+
+        self.history_theta = deque(maxlen=self.n_samples_per_update)
+        self.history_R = deque(maxlen=self.n_samples_per_update)
+        self.history_s = deque(maxlen=self.n_samples_per_update)
+        self.history_phi_s = deque(maxlen=self.n_samples_per_update)
+
+        # Evolution path for covariance
+        self.pc = np.zeros(self.n_params)
+        # Evolution path for sigma
+        self.ps = np.zeros(self.n_params)
+
+        self.mu = self.n_samples_per_update // 2
+        self.ordered_weights = np.zeros(self.n_samples_per_update)
+        self.ordered_weights[:self.mu] = (
+            np.log(self.mu + 0.5) - np.log1p(np.arange(int(self.mu))))
+        self.ordered_weights /= np.sum(self.ordered_weights)
+        self.mueff = 1.0 / np.sum(self.ordered_weights ** 2)
+
+        self.c1 = 2.0 / ((self.n_total_dims + 1.3) ** 2 + self.mueff)
+        self.cmu = min(
```

Comment:
... because it did not work. Don't ask me how this happened. :D

I also don't know if lambda is the population size or the regularization parameter in this case. It seems like someone forgot to update the formula.